airflow/operators/singleton_operator.py

```diff
@@ -0,0 +1,109 @@
+from airflow.utils.decorators import apply_defaults
+from airflow.utils.state import State
+
+
+class SingletonOperator(BaseOperator, SkipMixin):
+    """
+    Allows a workflow to ensure that only single `unfinished` task is being
+    executed at all times while allowing airflow to schedule multiple dags to
+    execute other task instances that are not directly downstream of the task.
+
+    Unfinished task states:
+            NONE,
+            SCHEDULED,
+            QUEUED,
+            RUNNING,
+            UP_FOR_RETRY
+
+    All directly downstream tasks will be skipped.
+
+    :param allow_external_trigger: If this is set, execution is allowed
+        regardless of dag_run states.
+    :type allow_external_trigger: Boolean
+    """
+
+    ui_color = '#e9ffdb'  # nyanza
+
+    @apply_defaults
+    def __init__(self, allow_external_trigger=False, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.task_instance = None
+        self.dag_run = None
+        self.downstream_task_instances = None
+        self.downstream_task_instance_ids = None
+        self.allow_external_trigger = allow_external_trigger
+
+    def execute(self, context):
+        if self.allow_external_trigger and context['dag_run'].external_trigger:
+            self.log.info("Externally triggered dag_run: allowing execution.")
+            return State.SCHEDULED.upper()
+
+        self.task_instance = context['ti']
+        self.dag_run = context['dag_run']
+        self.downstream_task_instances = self.get_flat_relatives(upstream=False)
+        self.downstream_task_instance_ids = [ti.task_id for ti in
+                                             self.downstream_task_instances]
+
+        self.log.info("downstream_task_instance_ids: %s",
+                      repr(self.downstream_task_instance_ids))
+        active_dag_runs = self.get_other_active_dag_runs()
+        self.log.info("Total active_dag_runs: %d", len(active_dag_runs))
+        result = self._skip_or_schedule(active_dag_runs)
+        self.log.info("%s execution: %s",
+                      result, repr(self.downstream_task_instance_ids))
+        return result
+
+    def _skip_or_schedule(self, active_dag_runs):
+        for active_dag in active_dag_runs:
+            tasks, skip = self.unfinished_tasks(active_dag)
+            if skip:
+                self.log.info(
+                    "Found active_dag: %s with running tasks: %s, skipping...",
+                    repr(active_dag), repr(tasks))
+                self.skip(self.dag_run,
+                          self.task_instance.execution_date,
+                          self.downstream_task_instances)
+                # Short circuit
+                return State.SKIPPED.upper()
+        return State.SCHEDULED.upper()
+
+    @provide_session
+    def unfinished_tasks(self, dag_run=DagRun, session=None):
+        TI = TaskInstance
+        tasks = session.query(TaskInstance).filter(
```

Comment:
Since you alias `TI` in the line above, let's use `TI` instead of `TaskInstance` here.

Reply:
Sure.